### PR TITLE
Add utilities functions to check material availability in trees

### PR DESCRIPTION
**Pull request by helderlopes97**

### CODE REVIEW

File: rxn4chemistry/core.py [hunk 1]
```diff
@@ -1107,6 +1107,43 @@ def update_node_actions(
         )
         return response
 
+    @response_handling(success_status_code=200, on_success=default_on_success)
+    @ibm_rxn_api_limits
+    def start_reaction_properties_prediction_from_smiles(
```
Comment (drugilsberg) [inline rxn4chemistry/core.py:1112]:
```suggestion
    def predict_reaction_properties_from_smiles(
```

File: rxn4chemistry/core.py [hunk 2]
```diff
@@ -1107,6 +1107,43 @@ def update_node_actions(
...
+            self,
+            reactions: List[str],
+            ai_model: str = "atom-mapping-2020-08-20",
+    ) -> requests.models.Response:
+        """
+        Launch prediction with given reactions SMILES.
+
+        Args:
+            ai_model (str, optional): model release. Defaults to
```
Comment (drugilsberg) [inline rxn4chemistry/core.py:1121]:
```suggestion
            ai_model (str, optional): model flavour and release. Defaults to
```

File: rxn4chemistry/utilities.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            return commercially_available
+        except KeyError:
+            logger.warning("No information on commercial availability")
+            return False
+    else:
+        # process children recursively
+        for child in tree["children"]:
+            if not check_starting_material_availability(child):
+                return False
+    return tree
```
Comment (drugilsberg) [inline rxn4chemistry/utilities.py:32]:
```suggestion
    commercially_available = False
    children = tree.get("children", [])
    # node is a leaf
    if len(children) == 0:
        try:
            commercially_available = tree["isCommercial"]
        except KeyError:
            logger.warning("No information on commercial availability")
            commercially_available = False
    else:
        # process children recursively
        for child in tree["children"]:
            if not check_starting_material_availability(child):
                commercially_available = False
    return commercially_available
```

File: rxn4chemistry/utilities.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    Args:
+        results (dict): Contains retrosynthetic_paths with all retrosynthesis trees that make the path.
+
+    Returns:
+        dict: returns a dict with the sequence ids as keys and commercially availability in all tree nodes as value.
+    """
+    return [
+        {
+            tree[index]["sequenceId"]: check_starting_material_availability(tree)
```
Comment (drugilsberg) [inline rxn4chemistry/utilities.py:47]: not sure what index is doing here, I believe that at this point would be better to return a list where each element is:
`{"index": index, "sequence_id": tree["sequenceId"], "are_materials_available": check_starting_material_availability(tree)}`

**Review by drugilsberg — CHANGES_REQUESTED** [review]

**Comment by helderlopes97** [comment]
@cla-bot check

**Review by drugilsberg — CHANGES_REQUESTED** [review]

File: rxn4chemistry/core.py [hunk 5]
```diff
@@ -1107,6 +1107,43 @@ def update_node_actions(
         )
         return response
 
+    @response_handling(success_status_code=200, on_success=default_on_success)
```
[the remark this answers is not in the document]
Comment (helderlopes97) [reply rxn4chemistry/core.py:1110]: I commented out the code in `predict_reaction_properties_from_smiles` and instead raised a NotImplementedError.

**Review by drugilsberg — APPROVED** [review]

**Pull request merged**